fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/common/MappedPool.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * MappedPool.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.common;
+
+import com.apple.foundationdb.annotation.API;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * MappedPool Class Attempts to reuse objects organized by keys.
+ * This class can be used for objects that require thread safety but our expensive to create.
+ *
+ *  Examples include Ciphers and Compressors
+ *
+ * @param <K> key
+ * @param <V> value type to be pooled
+ * @param <E> exception that can be throw, must extend Exception
+ */
+@API(API.Status.EXPERIMENTAL)
+public class MappedPool<K, V, E extends Exception> {
+    protected static final int DEFAULT_POOL_SIZE = 64;
+    protected final Function<K, Queue<V>> createQueue = k -> new ArrayBlockingQueue<>(DEFAULT_POOL_SIZE);
```

Comment:
Is there some weird pessimal case here where this actually makes things worse? In particular, I'm trying to think if there are some use cases where this, like, turns an average object from something that generally lives in the "short lived" JVM memory to something that graduates to the older generation before being dereferenced. Maybe something where, like, there are "bursts" of activity to the thing being cached? I'm not sure, though.

(This was commented on an earlier version of the code, but I'm copying it here to avoid having the comment linger on outdated code.)

Reply:
There are definitely pessimistic cases where this structure is not beneficial (cheap objects, cache coherency).  I do think this is a structure that is good to have in the toolkit for non-thread safe objects that are expensive to repetitively create.

Reply:
I was more trying to think if even for the more expensive objects that this is designed for, if certain use patterns wind up causing objects to live longer and therefore make garbage collection more expensive (even if the total number of object creations go down). I...think there are contrived use cases where that can be the case, but I'm not sure exactly how contrived (or if I'm right that those exist in general).

Reply:
@alecgrieser What is the ask?

Reply:
I'm not sure there's a concrete ask, here, so much as I just wanted to make sure that we've considered all the implications here of this change. Do you agree/disagree with my analysis? Or maybe does this need to be configurable (either queue size or usage in the serializer) if it may (but also may not) cause problems?

Reply:
@alecgrieser I understand G1 collection mechanisms and your analysis based on these concerns.  Since the Cypher sits in the Serde mechanism and the read your writes model hammers that portion (save and load record), I have considered your point.  I went ahead and added a constructor where you can supply the queue size.  I am less excited about a turn off or on configuration model (simply put, I would rather just reject the commit due to garbage collection "concerns").

Reply:
Okay, I see. That sounds fine, and I guess tweaking/tuning this is perhaps something we can look more into in the future. I don't think garbage collection concerns (to the extent that they are real and that this doesn't just make object creation better) aren't enough to reject the commit.